Step doubling for Implicit Euler Error Estimation

antequ opened the pull request:
This PR implements Implicit Euler with step doubling for error estimation. This change is very similar to error estimation for Velocity-Implicit Euler (VIE) Integrator, and if Evan has time, it would be great if he can review the high-level decisions made here. The VIE step-doubling error estimation is well-documented here, and the same documentation will apply for Implicit Euler: https://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_velocity_implicit_euler_integrator.html

Brief justification:
After running some tests with a pile of 3 objects and typical friction parameters, we found that step doubling works much better than implicit Trapezoid (IE+ITR) error estimation from the master branch. Under various accuracy settings (using a large max time step), IE+ITR performs very inconsistently, e.g. for one test, an accuracy of 0.1 takes 236423 derivative evaluations, while an accuracy of 0.099 takes only 82392. In most of these cases, step doubling is very consistent, achieving about 3x the performance of IE+ITR in the average case, 7-9x in the worst IE+ITR cases, and still 1.03x faster in the best IE+ITR case that I tested.
I suspect that the main reasons step doubling works better are:
1) IE with small steps is very likely to converge, after the big step already succeeded.
2) The same Jacobian can be used for both the big and small steps, and recomputed Jacobians are not wasted.

Some more important details about this change, which may be hard to understand from the code alone:

The major change from VIE's implementation is that Implicit Euler takes advantage of marking a Jacobian as "fresh" after a failed step, and never recomputing the Jacobian as it shrinks step sizes. The problem is, now, after it succeeds on the large step and the first small step, there is a rare (but nonzero) probability that it fails on the second small step, computing the Jacobian at a state that is not xt0. (In a stiff simulation of about 2000 steps with about 2000-4000 convergence-based failures, this scenario will occur maybe about ~20 times.) Based on some thought (and confirmed by experimentation), I've proposed a set of small changes that captures the performance gains of marking the Jacobian as "fresh" without triggering too much extra overhead. Note that this is much simpler than the first revision of the PR (I've removed a lot of caching logic which only improves performance by 2%, and replaced it with a TODO, as Sherm suggested):
1. After the Jacobian is computed, it is immediately marked as "fresh" (set_jacobian_is_fresh(true)).
2. Before starting the second step, we mark the Jacobian as "not fresh" (set_jacobian_is_fresh(false)). This way MaybeFreshenMatrices will recompute the Jacobian if the second small step is needed.
3. If the second small step actually fails, then we set a flag (failed_jacobian_is_from_second_small_step_) to indicate that in the next step, the Jacobian is NOT fresh, even though ImplicitIntegrator::DoStep() marks it as fresh.

This set of changes incurs minimal expenses during successful steps (only setting booleans), and it only forces the Jacobian to be recomputed again if the second small step actually fails. 

More justification:

Running Alejandro's difficult, 3-object pile-of-objects example, these are the resulting number of derivative evaluations:
https://docs.google.com/spreadsheets/d/12BSupktPFpbaTW2UbKaYLgUTAAF8upexcOAHXOjjJVI/edit?usp=sharing

All these rans ran it for 0.8 seconds, with maximum time step size 0.2.


Integrator: | IE (master, 6/5) |   | IE (PR #13224, 6/5) |  
-- | -- | -- | -- | --
  | Nsteps | Nderivatives | Nsteps | Nderivatives
Accuracy |   |   |   |  
0.5 | objects fly out |   | 1283 | 84686
0.25 | objects fly out |   | 1501 | 97803
0.2 | objects fly out |   | 1424 | 92206
0.15 | 2201 | 141691 | 1257 | 82922
0.13 | 1770 | 95981 | 1158 | 73327
0.11 | 1887 | 103152 | 1343 | 85775
0.1 | 2512 | 158336 | 1168 | 77289
0.099 | 2136 | 114818 | 1250 | 80553
0.094 | objects fly out |   | 1191 | 75051
0.06 | objects fly out |   | 1310 | 81758
0.03 | 2054 | 102249 | 1366 | 79836
0.01 | 21697 | 938885 | 1665 | 84754
0.005 | 29998 | 728335 | 2116 | 91576
0.002 | 51585 | 887838 | 3311 | 113554
0.001 | 224595 | 2094390 | 4360 | 129234



For almost all examples, the step doubling performed better (often much better) than trapezoid (master), and much more consistently.

Running the simple gripper example for two accuracies, these are the number of derivative evaluations:

bazel run //examples/simple_gripper:simple_gripper -- --mbp_discrete_update_period=0 --simulator_integration_scheme=implicit_euler --simulator_use_error_control=true --simulator_max_time_step=0.1 --simulator_accuracy=0.03

step doubling, 12318 derivative evals
trapezoid, 22757 derivative evals

bazel run //examples/simple_gripper:simple_gripper -- --mbp_discrete_update_period=0 --simulator_integration_scheme=implicit_euler --simulator_use_error_control=true --simulator_max_time_step=0.1 --simulator_accuracy=0.005

step doubling, 32185 derivative evals
trapezoid, 35555 derivative evals

When the accuracy is loose, step doubling performs much better.

Evan's suggestion of running all the tests in drake resulted in inconclusive results since runtime was dominated by the pydrake tests triggered by the simulator changes. I tried limiting it to just multibody and systems/analysis tests, and the new Implicit Euler appeared slightly better than the old implicit Euler (by 5% average runtime), but it's hard to say because those tests do not stress test this integrator, and timing produces results that vary per run.

-----------------------------------
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/13224)
<!-- Reviewable:end -->

jwnimmer-tri commented:
With my PR shepherd hat on -- over to +@edrumwri for right of first refusal on high level review, per @ante's note.

edrumwri commented:
@antequ I love seeing these improvements. Awesome!

I would love to be able to help with the review, but I'm already spread too thin. Here's a thought that might help though:

Many moons ago, we looked at replacing the RK3 integrator with the BS3 integrator (Matlab's default coarse accuracy integrator) as Simulator's default integrator because no one had any reason to believe that RK3 was performant. Certain integrators acquire a fanbase (e.g., aerospace engineers apparently love RK Merson) and RK3 didn't have much of one. To evaluate whether this decision would be worthwhile, we ran all of the bazel tests in release mode under both RK3 and BS3 (accounting for statistical noise, of course). RK3 performed markedly better.

I'd suggest doing something similar. Make IE+ITR the default integrator and run the tests, then compare against making IE+step doubling the default integrator. That would give some good evidence to argue which should be the selected default error estimation approach for IE.     

-@edrumwri

antequ commented:
Thanks for the suggestion, Evan! I assume that by "performed better" you meant timing the tests. I'll try that.

antequ commented:
___
*[systems/analysis/implicit_integrator.cc, line 417 at r2](https://reviewable.io/reviews/robotlocomotion/drake/13224#-M8W-52HA11xnyScQ6IJ:-M8WPM4e5Q6L6Pp3APV9:b-n2fzxr) ([raw file](https://github.com/robotlocomotion/drake/blob/8917351fe056daf697a0580e5982f150a5d1f66e/systems/analysis/implicit_integrator.cc#L417)):*
<details><summary><i>Previously, antequ (Ante Qu) wrote…</i></summary><blockquote>

That's not true. This part is assuming trial=1 failed after a bunch of successful steps. The Jacobian and the iteration matrix are very old. I added comments clarifying this in the VIE's version of this function, because you asked for this during the VIE PR; I can add them here.
</blockquote></details>

^That's the most likely use of this trial.

But yes, in the case where we just changed the step size, we DO do this twice, as you said. It's inefficient, and i can add an issue. We just assume it's rare enough, and factorizing is cheap enough, that it doesn't matter.


<!-- Sent from Reviewable.io -->
